File: ocaml/lib/context.ml
type t = { mutable pointer : unit Ctypes.ptr }

let create () =
  let ptr = Bindings.extism_context_new () in
  let t = { pointer = ptr } in
  Gc.finalise (fun { pointer } -> Bindings.extism_context_free pointer) t;
  t

let free ctx =
  let () = Bindings.extism_context_free ctx.pointer in
  ctx.pointer <- Ctypes.null

let reset ctx = Bindings.extism_context_reset ctx.pointer

let%test "test context" =
  let ctx = create () in
  reset ctx;
  free ctx;
  true

File: zig/src/context.zig
const std = @import("std");
const c = @import("ffi.zig");

const Self = @This();

mutex: std.Thread.Mutex,
ctx: *c.ExtismContext,

// We have to use this until ziglang/zig#2647 is resolved.
error_info: ?[]const u8,

/// Creates a new context, it should be freed using `deinit`
pub fn init() Self {
    const new_ctx = c.extism_context_new();
    return .{
        .mutex = .{},
        .ctx = new_ctx orelse unreachable,
        .error_info = null,
    };
}

// Free a context
pub fn deinit(self: Self) void {
    c.extism_context_free(self.ctx);
}

pub fn reset(self: *Self) void {
    self.mutex.lock();
    defer self.mutex.unlock();
    c.extism_context_reset(self.ctx);
}
